### ISSUE TITLE
Error when GPUs have double-digit IDs (device=10,11)

### ISSUE THREAD

**Issue by omarzd** [ultralytics/ultralytics#8481]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Multi-GPU

### Bug

[Issue 253](https://github.com/ultralytics/ultralytics/issues/253) is related. Although it doesn't mention double-digit IDs, I think the fix for this problem would be closely related to the fix for issue 253 was.

My system has 12 NVIDIA GeForce RTX 3070 devices. I can train on devices 0 to 9 successfully, all in parallel in fact. So the command `yolo train device=0,1,2,3,4,5,6,7,8,9` works as expected. However, as soon as I include device IDs 10 or 11, I get an Invalid CUDA device error as follows:

```
$ yolo train device=10
WARNING ⚠️ 'model' is missing. Using default 'model=yolov8n.pt'.
WARNING ⚠️ 'data' is missing. Using default 'data=coco8.yaml'.
Ultralytics YOLOv8.1.18 🚀 Python-3.8.0 torch-2.2.0+cu121 
Traceback (most recent call last):
  File "/home/omargui/.local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/cfg/__init__.py", line 568, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/engine/model.py", line 625, in train
    self.trainer = (trainer or self._smart_load("trainer"))(overrides=args, _callbacks=self.callbacks)
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/engine/trainer.py", line 100, in __init__
    self.device = select_device(self.args.device, self.args.batch)
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/utils/torch_utils.py", line 126, in select_device
    raise ValueError(
ValueError: Invalid CUDA 'device=10' requested. Use 'device=cpu' or pass valid CUDA device(s) if available, i.e. 'device=0' or 'device=0,1,2,3' for Multi-GPU.

torch.cuda.is_available(): True
torch.cuda.device_count(): 1
os.environ['CUDA_VISIBLE_DEVICES']: 0,1,2,3,4,5,6,7,8,9,10,11
```

I made sure that PyTorch can see all 12 devices using
```
import torch
print(torch.cuda.is_available())
print(torch.cuda.device_count())
for i in range(torch.cuda.device_count()):
    print(torch.cuda.get_device_name(i))
```
which outputs
```
TRUE
12
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
NVIDIA GeForce RTX 3070
```

I tried different ways of writing the string of devices like `device='10,11'` or `device=\'10,11\'`, but the problem persisted.  I believe there is a bug in how Ultralytics interprets the `device` parameter.

### Environment

```
$ yolo checks
Ultralytics YOLOv8.1.18 🚀 Python-3.8.0 torch-2.2.0+cu121 CUDA:0 (NVIDIA GeForce RTX 3070, 7974MiB)
Setup complete ✅ (4 CPUs, 23.4 GB RAM, 21.3/219.9 GB disk)

OS                  Linux-5.10.0-hiveos-x86_64-with-glibc2.27
Environment         Linux
Python              3.8.0
Install             pip
RAM                 23.40 GB
CPU                 Intel Core(TM) i5-6500 3.20GHz
CUDA                12.1

matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 10.2.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.2.0>=1.8.0
torchvision         ✅ 0.17.0>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@omarzd hey there! Thanks for the detailed report. It seems like the issue might be related to how the device IDs are parsed, especially when they hit double digits. A quick workaround could be setting the `CUDA_VISIBLE_DEVICES` environment variable before running your command. This way, you can map your physical GPU IDs to logical ones that start from 0. Here's how you can do it:

```bash
export CUDA_VISIBLE_DEVICES=10,11
yolo train device=0,1
```

This command sets GPUs with IDs 10 and 11 as the first and second devices for CUDA, and then you can refer to them as `device=0,1` in your training command. Give this a try and let us know if it helps! If the issue persists, it might indeed require a closer look into how the `device` parameter is interpreted. 🛠️

**Comment by omarzd** [comment]
@glenn-jocher Hi, Thanks for the quick reply. I just tried that. It ignored the `CUDA_VISIBLE_DEVICES ` variable and simply uses devices 0 and 1.

**Comment by glenn-jocher** [comment]
@omarzd for CLI arguments you can simply pass lists as python lists, i.e. `device=[0,1,10,11]`

**Comment by omarzd** [comment]
@glenn-jocher Thanks, but that didn't work either.

```
$ yolo train device=[10,11]
WARNING ⚠️ 'model' is missing. Using default 'model=yolov8n.pt'.
WARNING ⚠️ 'data' is missing. Using default 'data=coco8.yaml'.
Ultralytics YOLOv8.1.19 🚀 Python-3.8.0 torch-2.2.0+cu121
Traceback (most recent call last):
  File "/home/omargui/.local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/cfg/__init__.py", line 568, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/engine/model.py", line 625, in train
    self.trainer = (trainer or self._smart_load("trainer"))(overrides=args, _callbacks=self.callbacks)
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/engine/trainer.py", line 100, in __init__
    self.device = select_device(self.args.device, self.args.batch)
  File "/home/omargui/.local/lib/python3.8/site-packages/ultralytics/utils/torch_utils.py", line 126, in select_device
    raise ValueError(
ValueError: Invalid CUDA 'device=10,11' requested. Use 'device=cpu' or pass valid CUDA device(s) if available, i.e. 'device=0' or 'device=0,1,2,3' for Multi-GPU.

torch.cuda.is_available(): True
torch.cuda.device_count(): 2
os.environ['CUDA_VISIBLE_DEVICES']: 0,1,2,3,4,5,6,7,8,9,10,11
```

**Comment by glenn-jocher** [comment]
@omarzd It looks like the CLI isn't parsing the device list as expected. 🤔 For now, could you try setting the `CUDA_VISIBLE_DEVICES` environment variable directly in your shell before running the `yolo` command, like this:

```bash
export CUDA_VISIBLE_DEVICES=10,11
yolo train device=0,1
```

This approach manually sets which GPUs are visible to the process, effectively remapping `device=10,11` to `device=0,1`. We'll look into the parsing issue on our end. Thanks for bringing this to our attention!

**Comment by omarzd** [comment]
@glenn-jocher I have already tried this. It doesn't use devices 10 and 11. It uses devices 0 and 1.


> @glenn-jocher Hi, Thanks for the quick reply. I just tried that. It ignored the `CUDA_VISIBLE_DEVICES ` variable and simply uses devices 0 and 1.

**Comment by glenn-jocher** [comment]
@omarzd Thanks for giving that a try and for your patience! 🙏 It seems there's a misunderstanding with how `CUDA_VISIBLE_DEVICES` works. When you set `CUDA_VISIBLE_DEVICES=10,11`, CUDA remaps these to `device=0,1` for your process. So, when you then run `yolo train device=0,1`, it's actually using the GPUs you intended (10 and 11). If you're seeing unexpected behavior, it might be worth double-checking with a simple PyTorch script to list available devices after setting the environment variable. Here's a quick way to do that:

```python
import torch
print(torch.cuda.device_count())
for i in range(torch.cuda.device_count()):
    print(torch.cuda.get_device_name(i))
```

This should reflect the GPUs you've made available with `CUDA_VISIBLE_DEVICES`. If it still doesn't align, there might be another issue at play. Let's troubleshoot further from there!

**Comment by omarzd** [comment]
@glenn-jocher Yes it was clear from your first message what the expected behavior is from settings `CUDA_VISIBLE_DEVICES=10,11` is. I do see behavior other than what's expected. Let me be clear. I ran multiple experiments. Every single time, the value of the variable is completely ignored and the only thing taken into consideration is the `device` parameter in the `yolo` command. 

For instance, I set `CUDA_VISIBLE_DEVICES=10,11` or `2,4` or even unset it. When I run `yolo train device=0,1,2,3` or any indexes of devices below 10, I get exactly those devices running. Is my problem clear? The parsing of the `device` parameter value must have some bug in it when it comes to double-digit device IDs.

Furthermore, the script you shared behaves as expected. And since all my GPUs have the same name, it doesn't really provide more info than running a sample training. I have already shown sample output of that script in my bug report above.

Thank you for your support.

**Comment by omarzd** [comment]
I found the bug. It's in [`torch_utils.py`](https://github.com/ultralytics/ultralytics/blob/f8e681c2be251562633d424f4a1a1cc7da526bed/ultralytics/utils/torch_utils.py#L118)
In the line 
```
        if not (torch.cuda.is_available() and torch.cuda.device_count() >= len(device.replace(",", ""))):
          #raise error
```

Notice the `len(device.replace(",", ""))`. This is the wrong way to count the number of devices. It simply removes commas from the device string and checks the length of the resulting string, which will not work correctly for multi-digit device indices.

Here is the fix.

```
device_list = device.split(",")  # This will create a list of devices, e.g., "0,1,10" -> ['0', '1', '10']
num_devices_requested = len(device_list)

# Then you can compare num_devices_requested with the actual available device count
if not (torch.cuda.is_available() and torch.cuda.device_count() >= num_devices_requested):
    # ... raise error as before
```

**Comment by omarzd** [comment]
@glenn-jocher Please check [my pull request](https://github.com/ultralytics/ultralytics/pull/8502).

**Comment by glenn-jocher** [comment]
@omarzd Fantastic catch! 🌟 Your detailed analysis pinpointed the issue perfectly, and your proposed fix looks spot on. I've reviewed your pull request and it's exactly what we needed to address this bug. Thanks for taking the initiative to not only identify the problem but also to contribute a solution. Your effort is greatly appreciated by the entire YOLO community! 🙏 Let's get this merged and improve the experience for everyone using multi-digit device IDs.
